File: pilas/infijoTransformador.cpp
// USO DE LA LIBRERIA STACK PARA BREVE DEMOSTRACION DE TRANSFORMACION DE EXPRESIONES
// DE NOTACION INFIX A POSTFIX Y PREFIX
// NORMALMENTE SE USA UNA PILA PROPIA, AQUI USAMOS LA LIBRERIA STACK DE C++

#include <iostream>
#include <string>
#include <stack>
#include <algorithm>

using namespace std;

// Define la prioridad de los operadores.
int prioridad(char operador) {
    switch (operador) {
        case '+':
        case '-':
            return 1;
        case '*':
        case '/':
            return 2;
        case '^':
            return 3;
        default:
            return 0;
    }
}

// Convierte una expresión de notación infija a postfija.
string infijoAPostfijo(const string& expr) {
    stack<char> pila;
    string postfijo = "";

    for (char c : expr) {
        if (isalnum(c)) {
            postfijo += c;
        } else if (c == '(') {
            pila.push(c);
        } else if (c == ')') {
            while (!pila.empty() && pila.top() != '(') {
                postfijo += pila.top();
                pila.pop();
            }
            if (!pila.empty() && pila.top() == '(') {
                pila.pop();
            }
        } else { // Es un operador
            while (!pila.empty() && prioridad(c) <= prioridad(pila.top()) && pila.top() != '(') {
                postfijo += pila.top();
                pila.pop();
            }
            pila.push(c);
        }
    }

    while (!pila.empty()) {
        postfijo += pila.top();
        pila.pop();
    }

    return postfijo;
}

// Convierte una expresión infija a prefija.
string infijoAPrefijo(const string& expr) {
    string exprInvertida = "";
    for (int i = expr.length() - 1; i >= 0; --i) {
        char c = expr[i];
        if (c == '(') {
            exprInvertida += ')';
        } else if (c == ')') {
            exprInvertida += '(';
        } else {
            exprInvertida += c;
        }
    }

    string postfijoInvertido = infijoAPostfijo(exprInvertida);

    reverse(postfijoInvertido.begin(), postfijoInvertido.end());
    
    return postfijoInvertido;
}

int main() {
    string expresion1 = "A+B-C";
    cout << "Expresion infija: " << expresion1 << endl;
    cout << "Expresion postfija: " << infijoAPostfijo(expresion1) << endl;
    cout << "Expresion prefija: " << infijoAPrefijo(expresion1) << endl;

    string expresion2 = "(A+B)*C";
    cout << "\nExpresion infija: " << expresion2 << endl;
    cout << "Expresion postfija: " << infijoAPostfijo(expresion2) << endl;
    cout << "Expresion prefija: " << infijoAPrefijo(expresion2) << endl;

    string expresion3 = "(A*B)-(C+D)";
    cout << "\nExpresion infija: " << expresion3 << endl;
    cout << "Expresion postfija: " << infijoAPostfijo(expresion3) << endl;
    cout << "Expresion prefija: " << infijoAPrefijo(expresion3) << endl;

    return 0;
}

File: pilas/infijoTransformador.py
class Pila:
    def __init__(self):
        self.items = []

    def esta_vacia(self):
        return not self.items

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if self.esta_vacia():
            raise Exception("La pila está vacía.")
        return self.items.pop()

    def peek(self):
        if self.esta_vacia():
            raise Exception("La pila está vacía.")
        return self.items[-1]

def prioridad(operador):
    if operador in ['+', '-']:
        return 1
    if operador in ['*', '/']:
        return 2
    if operador in ['^']:
        return 3
    return 0

def infijo_a_postfijo(expr):
    pila = Pila()
    postfijo = []

    for c in expr:
        if c.isalnum():
            postfijo.append(c)
        elif c == '(':
            pila.push(c)
        elif c == ')':
            while not pila.esta_vacia() and pila.peek() != '(':
                postfijo.append(pila.pop())
            if not pila.esta_vacia() and pila.peek() == '(':
                pila.pop()
        else:
            while not pila.esta_vacia() and prioridad(c) <= prioridad(pila.peek()) and pila.peek() != '(':
                postfijo.append(pila.pop())
            pila.push(c)

    while not pila.esta_vacia():
        postfijo.append(pila.pop())

    return "".join(postfijo)

def infijo_a_prefijo(expr):
    expr_invertida = ""
    for c in expr[::-1]:
        if c == '(':
            expr_invertida += ')'
        elif c == ')':
            expr_invertida += '('
        else:
            expr_invertida += c
    
    postfijo_invertido = infijo_a_postfijo(expr_invertida)
    
    return postfijo_invertido[::-1]

if __name__ == "__main__":
    expresion1 = "A+B-C"
    print(f"Expresión infija: {expresion1}")
    print(f"Expresión postfija: {infijo_a_postfijo(expresion1)}")
    print(f"Expresión prefija: {infijo_a_prefijo(expresion1)}")

    expresion2 = "(A+B)*C"
    print(f"\nExpresión infija: {expresion2}")
    print(f"Expresión postfija: {infijo_a_postfijo(expresion2)}")
    print(f"Expresión prefija: {infijo_a_prefijo(expresion2)}")

    expresion3 = "(A*B)-(C+D)"
    print(f"\nExpresión infija: {expresion3}")
    print(f"Expresión postfija: {infijo_a_postfijo(expresion3)}")
    print(f"Expresión prefija: {infijo_a_prefijo(expresion3)}")